### DISCUSSION TITLE
WebApp extension build procedure

### DISCUSSION THREAD

**Discussion by hmiyake** [DIRACGrid/DIRAC#6367] [Q&A] [answered]
Hello,

Let me confirm if my procedure to build WebApp extension is valid.

Basically I follow the instruction given by
https://github.com/DIRACGrid/dirac-webapp-packaging

Essential files, setup.py, setup.cfg, pyproject.toml, MANIFEST.in, are referring to
https://gitlab.cern.ch/lhcb-dirac/LHCbWebDIRAC

Then
- prepare source code repository managed by git
- tag the repository obeying PEP440
- run `python -m build` to get wheel package

Is it correct?

In fact our WebApp extension code is made by special way (not managed by git), then setuptools_scm does not work.
So currently I need to remove the tool from pyproject.toml.

At the last BiLD, what I mentioned was tarball creation by `python setup.py sdist`, but it resulted in wheel build during deployment.
Singularity needs to run on DIRAC node but could crash depending on environment, then I wonder if it is correct way...yes it's wrong.

**Comment by fstagni** [comment]
While not answering your question, did you also follow https://github.com/DIRACGrid/DIRAC/wiki/How-to-migrate-to-new-_WebHandler-class ?

**Reply by hmiyake** [reply]
Wow, I haven't realized that! Thank you for notification. We'll apply this recipe at 8.0 migration as well as other big changes described in [the update instruction to 8.0](https://github.com/DIRACGrid/DIRAC/wiki/DIRAC-8.0)

**Comment by chrisburr** [comment, marked as answer]
Following up from the BiLD: 

* The steps above seem reasonable
* For the dependencies the only need to be modified when changing between release series of DIRAC (i.e. 7.2 -> 7.3 -> 8.0 -> 8.1):
    * `LHCbWebDIRAC` depends on `WebAppDIRAC >=A.B.C,<A.(B+1)`
    * `LHCbWebDIRAC` depends on `LHCbDIRAC >=X.Y.Z,<Z.(Y+1)`
* We currently only support installation from the offical PyPI repository at pypi.org though @hmiyake mentioned this isn't acceptable for them.
    * The easiest solution would be to look at exactly what the policy is and see if it can be changed
    * If it can't be, do you already have a PyPI mirror available? (Using something like artifactory or nexus)
    * If you don't have a mirror you will need to set one up
    * The SystemAdministrator/Pilot can be relatively easily extended to support non-default PyPI servers and it will be much simplier than writing your own installer.

**Reply by hmiyake** [reply]
Thank you for your feedback!

Currently we use ordinal http server to provide our package (i.e. no valid mirror but just providing our own package)
My command line is something like 
pip install BelleDIRAC==5.6.0a1 --extra-index-url=http://XXX.YYY.ZZZ/packages --trusted-host XXX.YYY.ZZZ
Or if direct file specification like
pip install /cvmfs/XXX.YYY.ZZZ/.../BelleDIRAC-1.0.0-py3-none-any.whl
is accepted, it is also appreciated

**Reply by chrisburr** [reply]
In that case I think I've misunderstood what the limitation is.

Is it okay to use PyPI and the only issue is with uploading BelleDIRAC to it?

**Reply by hmiyake** [reply]
Right, we can take necessary packages from PyPI, including DIRAC, but don't want to publish BelleDIRAC
